### PR TITLE
Dashboard Toast Actions Update

### PR DESCRIPTION
**Pull request by MVarshini**
PBENCH-835

Dashboard CLEAR_TOAST action

Removed setValue() function as it mutates Redux store and updated the toast actions accordingly

### CODE REVIEW

**Comment by MVarshini** [comment]
Ordering of imports is happening automatically as I save the file because of the editor extension installed.

**Review by portante — CHANGES_REQUESTED** [review]
A couple things to consider before getting into the review.

The second commit is a "merge" commit.  Can you use `git rebase` instead of `git merge`?  See https://opensource.com/article/20/11/multiple-git-repositories for details on how to setup your upstream branch so it will work cleanly.

We really need to include unit tests and/or functional tests along with code changes.  Can you work to add them to this PR?

Thanks!

**Comment by portante** [comment]
Moved to `Draft` mode until we can add testing to this PR.

**Comment by dbutenhof** [comment]
> We really need to include unit tests and/or functional tests along with code changes. Can you work to add them to this PR?

Shubham is still struggling with the basic unit test infrastructure and mocks, so I'm not sure that's practical until we get that working... unfortunately. 😦

**Comment by portante** [comment]
> > We really need to include unit tests and/or functional tests along with code changes. Can you work to add them to this PR?
> 
> Shubham is still struggling with the basic unit test infrastructure and mocks, so I'm not sure that's practical until we get that working... unfortunately. 😦

Perhaps all-hands to help Shubham first then ...

**Comment by webbnh** [comment]
> > > We really need to include unit tests and/or functional tests along with code changes. Can you work to add them to this PR?
> > 
> > 
> > Shubham is still struggling with the basic unit test infrastructure and mocks, so I'm not sure that's practical until we get that working... unfortunately. 😦
> 
> Perhaps all-hands to help Shubham first then ...

Yeah.  Testing is foundational, so we need it to be as easy as possible to add and run tests.  And, we probably shouldn't be waiting on someone who has other concerns, so and all-hands jam session would be appropriate.

**Comment by pravins** [comment]
@shubham-html-css-js mentioned, he has worked on it and need further help from @dbutenhof . He might connect with Dave soon.

**Comment by dbutenhof** [comment]
> @shubham-html-css-js mentioned, he has worked on it and need further help from @dbutenhof . He might connect with Dave soon.

To be clear: I've worked with Shubham several times on this, but I'm afraid the problem he's hitting is beyond my depth in React and jest. At some point the Javascript code is trying to access localhost port 80. There's no Pbench dashboard code on the stack, so I suspect this is happening during Reach initialization before his mock is registered.

I know there is a special directory from which the infrastructure will load mocks early in initialization, but I also know that even that wasn't sufficient (in the old dashboard structure) to instantiate mocks early enough to catch the `GET /endpoints` in `index.html` and we never found a solution for that. (Varshini's solution of deferring the endpoints call into Redux is probably the best: I'd been hesitant to approach that in the old dashboard because Gurbir had spent a lot of time hoisting the configuration file load *out* of Redux.) 

Anyway, I've got no idea if this problem Shubham is hitting is related, and from the debugging information available it's hard to get a handle on exactly where this is happening. (We're levels deep in `node_module` stacks without clear names.)

File: dashboard/src/actions/toastActions.js
Comment (webbnh) [inline dashboard/src/actions/toastActions.js:40]: As I understand it, `alerts` is a reference to part of the Redux state, and we're not supposed to mutate that; yet, line 35 is doing so.  It's true that the mutation doesn't last for long before we update the state, but I think we're still technically outside the rules.

So, the code should probably do something more like this:
```
    const obj = {
      variant: variant,
      title: title,
      message: message,
      key: uid(),
    };
    const alerts = [...getState().toastReducer.alerts, obj]

    dispatch({
      type: TYPES.SHOW_TOAST,
      payload: alerts,
    });
```
which creates the object that we want to add to the alerts list, then creates a copy of the alerts list with the object added at the end (leaving the Redux state unchanged), and then passes the copy to `dispatch()` to be used in the update of the state.

**Review by webbnh — CHANGES_REQUESTED** [review]
It looks to me like, with the proposed changes, the `CLEAR_TOAST` event is still clearing the whole Redux state and replacing it with just the provided Toast actions...which seems...undesirable...but perhaps I don't quite understand exactly what the `CLEAR_TOAST` event is supposed to do.

Also, it looks like the new `show_toast` function is briefly mutating the Redux state.

Finally, I'm wondering whether we actually need or want the `CLEAR_TOAST` event -- it seems like we could use the `SHOW_TOAST` event instead (in which case, we might want to rename the event to something more like `UPDATE_TOAST`...).

**Comment by dbutenhof** [comment]
Intriguing: it looks like the `npx eslint` stage is failing over errors in the `node_modules/railroad_diagrams` packages, despite the `"src/**"` argument:

```
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:7:6: E225 missing whitespace around operator
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:12:1: E302 expected 2 blank lines, found 1
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:15:1: E302 expected 2 blank lines, found 1
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:26:1: E303 too many blank lines (3)
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:58:47: E231 missing whitespace after ','
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:91:5: E303 too many blank lines (2)
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:139:5: E303 too many blank lines (2)
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:204:9: F841 local variable 'last' is assigned to but never used
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:213:21: E131 continuation line unaligned for hanging indent
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:214:21: E131 continuation line unaligned for hanging indent
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:223:21: E131 continuation line unaligned for hanging indent
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:224:21: E131 continuation line unaligned for hanging indent
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:238:25: E131 continuation line unaligned for hanging indent
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:239:25: E131 continuation line unaligned for hanging indent
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:247:21: E131 continuation line unaligned for hanging indent
./dashboard/node_modules/railroad-diagrams/railroad_diagrams.py:248:21: E131 continuation line unaligned for hanging indent
```

**Comment by webbnh** [comment]
> it looks like the `npx eslint` stage is failing over errors in the `node_modules/railroad_diagrams` packages, despite the `"src/**"` argument

It's even more confusing that that -- according to the [eslint docs](https://eslint.org/docs/latest/user-guide/configuring/ignoring-code#the-eslintignore-file):
> In addition to any patterns in the `.eslintignore` file, ESLint always follows a couple of implicit ignore rules even if the `--no-ignore` flag is passed. The implicit rules are as follows:
> 
> • `node_modules/` is ignored.

Looking at the output in the log,
- the current directory should have been `dashboard`, not its parent;
- we requested that eslint check only the `src` subdirectory;
- we didn't ask for the `node_modules` subdirectory to be checked; and,
- the `node_modules` subdirectory should have been ignored implicitly.

So, I can only conclude that I really don't understand the situation.  😖

**Comment by webbnh** [comment]
I do have two tidbits to offer (not that they are actually useful in addressing this situation)....

The first is that the _documentation_ for `eslint` indicates that command line option switches are supposed to precede the path(s) to be linted, but we have the path-glob first.

The second is that we have an `"eslintConfig"` entry in our `package.json` file, which (according to the docs) will be ignored because we also have an `.eslintrc.json` file which takes precedence.  (So, we should remove the entry from our `package.json` file and consider merging it into our `.eslintrc.json` file.)

But, I don't see how either of these would trigger the warnings that we're seeing, here.  😕

**Review by webbnh — COMMENTED** [review]
Here's a quick observation on the new code you posted.

Also, I think I have fixed the lint problem in PR #3104 -- you should rebase your branch on that once it's merged.

File: dashboard/src/reducers/toastReducer.js
Comment (MVarshini) [reply dashboard/src/reducers/toastReducer.js]: Updated the `CLEAR_TOAST` action case in the reducer.

File: dashboard/src/actions/toastActions.js
Comment (MVarshini) [reply dashboard/src/actions/toastActions.js]: Yes, it should!

File: dashboard/src/modules/components/ToastNotificationComponent/index.jsx
Comment (webbnh) [inline dashboard/src/modules/components/ToastNotificationComponent/index.jsx:22]: Here's another place where you can get rid of the block and the `return` statement, if you like.

File: dashboard/src/store/mockStore.js
Comment (webbnh) [inline dashboard/src/store/mockStore.js:10]: I think the blank line at line 9 belongs between lines 7 & 8, so that the comment at line 8 is "more attached" to line 10.

**Review by webbnh — COMMENTED** [review]
I don't understand one of the tests.  (And, I've a couple of nits for you.)

**Review by webbnh — CHANGES_REQUESTED** [review]
I don't think we've entirely resolved the issue with the `CLEAR_TOAST` reducer.  Also, there are new build warnings, and I think I spotted a pointless test.  We should try to resolve these items before merging.

**Comment by webbnh** [comment]
And, hopefully, rebasing this branch on `main` will resolve the current lint failures (but...I cannot promise that it won't find new ones... 🤷).

File: dashboard/src/reducers/toastReducer.js
Comment (webbnh) [reply dashboard/src/reducers/toastReducer.js]: This conversation [resumes below](https://github.com/distributed-system-analysis/pbench/pull/3021#discussion_r1042344032) (so I'm going to resolve it here).

File: dashboard/src/modules/components/ToastNotificationComponent/index.jsx
Comment (MVarshini) [reply dashboard/src/modules/components/ToastNotificationComponent/index.jsx]: Updated!

File: dashboard/src/store/mockStore.js
Comment (MVarshini) [reply dashboard/src/store/mockStore.js]: Got it!

File: dashboard/src/App.test.js
Comment (MVarshini) [reply dashboard/src/App.test.js:17]: It's on hold for now!

File: dashboard/src/setupTests.js
Comment (MVarshini) [reply dashboard/src/setupTests.js]: To use `enzyme` with React 17 an unofficial adapter `enzyme-adapter-react-17-updated` is required which is causing dependency conflicts. So, removed `enzyme` and made sure nothing is affected.

File: dashboard/src/App.test.js
Comment (webbnh) [reply dashboard/src/App.test.js:17]: This presents a dilemma:  as a rule, we should be resistant to having "placeholders" in the code, particularly ones which cost us execution time like this.  They are clutter, and they tend to "rot" and cause other hassles.

My recommendation is that, unless you're intending to add something to this test for this PR, just go ahead and remove the whole function.  It will be easy enough to add back in when you are ready for it to actually do something.  If you really want to keep it in the code, then please add a comment explaining the situation, so that future developers will know why the no-op code is there and so that they will have some context when deciding when and whether they can delete it.

If you're simply concerned about it being forgotten, then create an issue for it in Jira and remove it from the code.

**Review by webbnh — COMMENTED** [review]
It looks good to me, now.

The only outstanding issue is the [empty test](https://github.com/distributed-system-analysis/pbench/pull/3021#discussion_r1044622455), although the change to use a string literal instead of a symbolic constant for the navigation target seems like it should be probably changed back.

Other than those, I just have a nit and a couple of suggestions.

File: dashboard/package.json
Comment (dbutenhof) [inline dashboard/package.json:20]: Can we get rid of the trailing whitespace here?

File: dashboard/package.json
Comment (dbutenhof) [inline dashboard/package.json:78]: Trailing whitespace here, also

File: dashboard/src/store/mockStore.js
Comment (dbutenhof) [inline dashboard/src/store/mockStore.js:12]: I suspect this name should be more specific, like `mockToastState`, since I expect there will be more "store" mocks in the future? I think a better name would also make the test case a little easier to read since there's less room to wonder what's really there. 😄

Comment (MVarshini) [reply dashboard/src/store/mockStore.js:12]: The same store can be used for different reducers.

File: dashboard/src/App.test.js
Comment (MVarshini) [reply dashboard/src/App.test.js:17]: Removed the placeholder test case. Created [PBENCH-1038](https://issues.redhat.com/browse/PBENCH-1038) to track

File: dashboard/package.json
Comment (MVarshini) [reply dashboard/package.json]: Removed the blank line

File: dashboard/src/actions/authActions.js
Comment (MVarshini) [reply dashboard/src/actions/authActions.js]: Updated

File: dashboard/src/reducers/toastReducer.js
Comment (MVarshini) [reply dashboard/src/reducers/toastReducer.js]: Done!

File: dashboard/src/actions/toastActions.js
Comment (MVarshini) [reply dashboard/src/actions/toastActions.js]: Updated!

File: dashboard/package.json
Comment (MVarshini) [reply dashboard/package.json]: Removed the trailing whitespace

File: dashboard/package.json
Comment (MVarshini) [reply dashboard/package.json]: Done!

**Review by dbutenhof — APPROVED** [review]

**Review by webbnh — APPROVED** [review]
👍

**Pull request merged**
